### PR TITLE
ci(aio) deployment mode features

### LINKED ISSUE #18189: create next.angular.io + add "next" entry to the version dropdown
## Current behavior

Master branch docs are being published to aio-staging.firebaseapp.com but we don't advertise this anywhere, so nobody can easily find the documentation for the current beta/rc release.

## Expected behavior

- [ ] we should alias aio-staging.firebaseapp.com to next.angular.io
- [ ] add an entry called "next" into the version dropdown that will redirect to next.angular.io.

### LINKED ISSUE #18287: archive/stable/next modes for angular.io
With Angular v5 coming in September, we need to have a solid plan for the docs and their versioning. v4 is LTS so we must keep the docs for v4 around at an easy to find location.

After speaking with @petebacondarwin and @juleskremer here is our proposal:

- Add a "mode" feature to aio app
  - there would be three recognized modes:
    - stable (used by angular.io)
    - next (used by next.angular.io)
    - archive (used by vXX.angulario, e.g. v4.angular.io)
  - the mode feature would control four things within the app:
    - theme (colors used within the header and footer of the app, maybe more? tbd)
      - stable mode would use the current angular.io palette
      - next mode would use slightly different palette (more bright? maybe red?)
      - archive mode would use gray color palette
    -  redirection of all marketing urls to /docs in archive and next mode
       - the marketing pages are irrelevant in the next and archived mode and it's more appropriate to just redirect the visitor to the actual documentation)
       - the redirection should happen via Location service (using history.replaceState) so that we don't interfere with the critical path for TTI and don't need to maintain special firebase hosting config for various environments.
- Mode activation
  - the correct mode should be activated either as a post-build/pre-deploy time or at bootstrap time 
    - in both cases we understand the target environment
    - activating the mode during the pre-deploy time would allow us to better control robots.txt to ban crawling of next and archive app - something we don't do today.
  - for local development assume the default to be the "stable" mode
  - allow mode to be changed using query parameter for development / debugging
- Add robots.txt to all archived and the next deployments to prevent crawling of these
- Adjust the version picker to contain info about next/stable and archived versions and select the one appropriately
  - two new entries should be added to version picker: next and stable
  - all archived versions should be added to the picker as we add them
  - the currently active version should be selected in the dropdown by default
    - if the current version is also stable or next, the version number should be appended to the item in the dropdown menu
    - example dropdown menu content:
       ```
       - next
       - [stable (v4.3.2)]
       - v2
       ```

### PR DESCRIPTION
**Pull request by petebacondarwin**
Closes #18287 
Closes #18189

- [x] deployment to firebase
- [x] mode activation (via environment const or URL query param)
- [x] `aio-shell` element css class (`mode-{archive/latest/next}`)
- [x] info banner for archive mode
- [x] add robots.txt for archive mode
- [x] redirect marketing pages
- [x] adjust version dropdown menu
- [x] CSS style themes

### CODE REVIEW

File: aio/src/app/app.component.html [hunk 1]
```diff
@@ -21,8 +21,8 @@
     <aio-nav-menu *ngIf="!isSideBySide" [nodes]="topMenuNarrowNodes" [currentNode]="currentNodes?.TopBarNarrow" [isWide]="false"></aio-nav-menu>
     <aio-nav-menu [nodes]="sideNavNodes" [currentNode]="currentNodes?.SideNav" [isWide]="isSideBySide"></aio-nav-menu>
 
-    <div class="doc-version" title="Angular docs version {{currentDocVersion?.title}}">
```
Comment (gkalpak) [inline aio/src/app/app.component.html]: OOC, why?

File: aio/src/app/app.component.ts [hunk 2]
```diff
@@ -129,6 +129,10 @@ export class AppComponent implements OnInit {
     });
 
     this.locationService.currentPath.subscribe(path => {
+      // Redirect to docs if we are in archive mode and not hitting a docs page (i.e. a marketing page)
```
Comment (gkalpak) [inline aio/src/app/app.component.ts]: _we are in archive mode_ --> _we are not in stable mode_/_we are in archive or next mode_

File: aio/src/app/app.component.spec.ts [hunk 3]
```diff
@@ -765,6 +765,33 @@ describe('AppComponent', () => {
...
+
+        createTestingModule('tutorial/toh-pt1', 'archive');
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+
+        createTestingModule('docs', 'archive');
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+      });
+    })
```
Comment (gkalpak) [inline aio/src/app/app.component.spec.ts]: Missing `;` 🙀

File: aio/src/app/app.component.spec.ts [hunk 4]
```diff
@@ -765,6 +765,33 @@ describe('AppComponent', () => {
...
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+
+        createTestingModule('tutorial/toh-pt1', 'archive');
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+
+        createTestingModule('docs', 'archive');
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+      });
```
Comment (gkalpak) [inline aio/src/app/app.component.spec.ts]: No tests for other modes 😞

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: aio/.angular-cli.json [hunk 5]
```diff
@@ -31,8 +31,9 @@
       "environmentSource": "environments/environment.ts",
       "environments": {
         "dev": "environments/environment.ts",
-        "stage": "environments/environment.stage.ts",
-        "prod": "environments/environment.prod.ts"
```
Comment (petebacondarwin) [inline aio/.angular-cli.json]: OK

File: aio/src/environments/environment.ts [hunk 6]
```diff
@@ -14,5 +14,6 @@ import 'core-js/es7/reflect';
 
 export const environment = {
   gaId: 'UA-8594346-26', // Staging site
-  production: false
+  production: false,
+  mode: 'stable'
```
Comment (petebacondarwin) [inline aio/src/environments/environment.ts]: I think we agreed that dev env would just use this `gaId` since we can filter out `localhost` requests??  @wardbell or @IgorMinar  might know more.

File: aio/scripts/deploy-to-firebase.sh [hunk 7]
```diff
@@ -3,34 +3,42 @@
...
     ;;
+  archive)
+    readonly majorVersion=${TRAVIS_BRANCH%%.*}
+    readonly projectId=angular-io-${majorVersion}
+    readonly deployedUrl=https://v${majorVersion}.angular.io/
+    readonly firebaseToken=$FIREBASE_TOKEN
+    ;;
   *)
-    echo "Unknown deployment environment ('$deployEnv'). Expected 'staging' or 'production'."
+    echo "Unknown deployment environment ('$deployEnv'). Expected 'next', 'stable' or 'archive'."
```
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: OK

File: scripts/ci/deploy.sh [hunk 8]
```diff
@@ -49,21 +49,17 @@ case ${CI_MODE} in
         exit 0
```
Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: :-( you're right. How about we try to deploy for any build that is at the head of a branch?
The actual firebase deploy will fail because there will be no matching FB project.

For example, building `my-temp-branch` on `angular/angular` would result in `deploy-to-firebase.sh` attempting to deploy to the FB project called `angular-io-my-temp-branch` at the url `https://vmy-temp-branch.angular.io/`

Or is this too hacky?

Otherwise we have to somehow work out if the branch is a valid archive branch or not...

File: aio/src/app/app.component.html (hunk 1 above)
Comment (petebacondarwin) [inline aio/src/app/app.component.html]: This just seems wrong. The `div` is a container that holds the drop-down to select versions. Why should it have a `title` that describes just one of the versions?
If anything I would expect it to have a title that said something like `"Select a version of Angular"`.

Moreover, because there were no other `title` attributes, further down, it meant that this tooltip pops up when you hover over any of the individual versions in the drop-down.

image

image

File: aio/src/app/app.component.spec.ts (hunk 4 above)
Comment (petebacondarwin) [inline aio/src/app/app.component.spec.ts]: gotcha!

File: aio/src/app/app.component.spec.ts (hunk 3 above)
Comment (petebacondarwin) [inline aio/src/app/app.component.spec.ts]: OMG!

File: scripts/ci/deploy.sh (hunk 8 above)
Comment (gkalpak) [inline scripts/ci/deploy.sh]: But archive branches will not deploy to a project matchig their name exactly, would they? Because we only care about the major version.
We need to somehow detect if a branch is an archive branch and even further detect if it is the latest branch for that major version. E.g. we don't want to deploy to `v4.angular.io` from `4.2.x` once `4.3.x` has been released.

Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: Agh! Yes we don't want to deploy from 4.2.x at all once 4.3.x is released...
What to do ...

File: aio/src/app/app.component.html (hunk 1 above)
Comment (IgorMinar) [inline aio/src/app/app.component.html]: was this done for (wrong) accessibility reasons? AFAIK the title attribute is not useful in this context.

File: scripts/ci/deploy.sh (hunk 8 above)
Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: So instead when we have an archive release, we search github for the most recent minor release for that archive's major version and only deploy if that matches the current branch, right?

**Review by IgorMinar — COMMENTED** [review]
There is no preview link - I guess because of the recent CI issues.

Overall this looks very good. Posting a few minor comments. Let's wait for the preview before merging this in.

File: scripts/ci/deploy.sh (hunk 8 above)
Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: I updated the code for this... PTAL

File: aio/src/app/app.component.html (hunk 1 above)
Comment (petebacondarwin) [inline aio/src/app/app.component.html]: I guess that was the case.

File: aio/src/app/layout/mode-banner/mode-banner.component.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Component, Input } from '@angular/core';
+import { VersionInfo } from 'app/navigation/navigation.service';
+
+@Component({
+  selector: 'aio-mode-banner',
+  template: `
+  <div *ngIf="mode == 'archive'" class="mode-banner">
+    This is the <strong>archived documentation for Angular v{{version?.major}}.</strong>
```
Comment (petebacondarwin) [inline aio/src/app/layout/mode-banner/mode-banner.component.ts]: I'll check

File: aio/src/extra-files/archive/robots.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+User-agent: *
+Disallow: /
```
Comment (petebacondarwin) [inline aio/src/extra-files/archive/robots.txt]: done

File: aio/src/app/shared/deployment.service.ts [hunk 11]
```diff
@@ -1,7 +1,17 @@
...
+ * Information about the deployment of this application.
+ */
 @Injectable()
 export class Deployment {
-  mode: string = environment.mode;
+  /**
+   * The deployment mode set from the environment provided at build time;
+   * or overridden by the `deploymode` query parameter: e.g. `...?deploymode=archive`
+   */
+  mode: string = this.location.search()['deploymode'] || environment.mode;
```
Comment (petebacondarwin) [inline aio/src/app/shared/deployment.service.ts]: I was worried that 'mode' would be too generic and might conflict with some future feature...
Since this is a debug thing, how about '_mode'?

Comment (petebacondarwin) [inline aio/src/app/shared/deployment.service.ts]: Actually let's just go with 'mode' - first come first served

File: aio/src/app/layout/mode-banner/mode-banner.component.ts (hunk 9 above)
Comment (petebacondarwin) [inline aio/src/app/layout/mode-banner/mode-banner.component.ts]: No, the version info is set via a subscription in the AppComponent: https://github.com/angular/angular/blob/205abe814017a48ec602db959dea37966292dd32/aio/src/app/app.component.ts#L176

**Comment by petebacondarwin** [comment]
Updated @IgorMinar - PTAL

File: aio/src/environments/environment.ts (hunk 6 above)
Comment (wardbell) [inline aio/src/environments/environment.ts]: What he said ^^^

**Comment by petebacondarwin** [comment]
Some examples:

Archive mode: https://pr18297-771a433.ngbuilds.io/tutorial?mode=archive

screen shot 2017-07-27 at 07 14 12

Next mode: https://pr18297-771a433.ngbuilds.io/tutorial?mode=archive

screen shot 2017-07-27 at 07 15 00

Redirection - clicking on this link will redirect you to the docs page: https://pr18297-771a433.ngbuilds.io/resources?mode=next. Note that the query string doesn't survive the redirection (which is not a problem in practice) but does mean that the mode disappears after the redirection in this case.

File: aio/package.json [hunk 12]
```diff
@@ -9,7 +9,7 @@
     "ng": "yarn check-env && ng",
     "start": "yarn check-env && ng serve",
     "prebuild": "yarn check-env && yarn setup",
-    "build": "ng build -prod -sm -vc=false",
+    "build": "ng build ng build --target=production --environment=stable -sm -vc=false",
```
Comment (petebacondarwin) [inline aio/package.json]: This change fixes @gkalpak comment https://github.com/angular/angular/pull/18297/files#r129411614

File: aio/src/app/app.component.spec.ts (hunk 4 above)
Comment (petebacondarwin) [inline aio/src/app/app.component.spec.ts]: added

File: aio/scripts/deploy-to-firebase.sh [hunk 13]
```diff
@@ -3,41 +3,77 @@
...
+# The deployment mode is computed based on the branch we are building
+if [ $TRAVIS_BRANCH == master ]; then
+  readonly deployEnv=next
+elif [ $TRAVIS_BRANCH == $STABLE_BRANCH ]; then
+  readonly deployEnv=stable
+else
+  readonly majorVersion=${TRAVIS_BRANCH%%.*}
+  readonly mostRecentMinorVersion=$(
+    # List the branches that start with the major version
+    git ls-remote https://github.com/angular/angular refs/heads/${majorVersion}.*.x |
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: Nit: Why not use `origin`? (We already use it to determine the lastest commit.)

File: aio/scripts/deploy-to-firebase.sh [hunk 14]
```diff
@@ -3,41 +3,77 @@
...
+    # Extract the version number
+    awk -F'/' '{print $3}' |
+    # Sort by the minor version
+    sort -t. -k 2,2n |
+    # Get the highest version
+    tail -n1)
+  if [ $TRAVIS_BRANCH == $mostRecentMinorVersion ]; then
+    readonly deployEnv=archive
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: This will still deploy `4.2.x` to `v4.angular.io` when `4.3.x` is the stable. Is it intentional?

File: aio/scripts/deploy-to-firebase.sh [hunk 15]
```diff
@@ -3,41 +3,77 @@
...
+    exit 0
+  fi
+fi
 
 case $deployEnv in
-  staging)
-    readonly buildEnv=stage
+  next)
     readonly projectId=aio-staging
     readonly deployedUrl=https://$projectId.firebaseapp.com/
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: Shouldn't this be `https://next.angular.io/`?

File: aio/scripts/deploy-to-firebase.sh [hunk 16]
```diff
@@ -3,41 +3,77 @@
...
 # Deploy
 (
   cd "`dirname $0`/.."
 
   # Build the app
-  yarn build -- --env=$buildEnv
+  yarn build -- --env=$deployEnv
+
+  # Include any mode-specific files
+  cp -rf src/extra-files/$buildEnv/. dist/
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: `$buildEnv` --> `$deployEnv`

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: aio/src/environments/environment.ts [hunk 17]
```diff
@@ -13,6 +13,7 @@ import 'core-js/es7/reflect';
 
 
 export const environment = {
-  gaId: 'UA-8594346-26', // Staging site
-  production: false
+  gaId: 'UA-8594346-26', // Staging id
```
Comment (petebacondarwin) [inline aio/src/environments/environment.ts]: OK - let's call it dev

File: aio/scripts/deploy-to-firebase.sh (hunk 13 above)
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: Because I was worried that `origin` might not exist on a CI server. But I guess, if we are using it elsewhere then it is less fragile for using with a different repo.

File: aio/scripts/deploy-to-firebase.sh (hunk 14 above)
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: While 4.3.x is stable, there will be no `angular-io-4` firebase project, so it would just fail.
Perhaps we want not to fail but just quietly not deploy...

Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: OK, so I'll change it so that we quietly fail if there is a branch with the same major but a higher minor version number - even if that higher branch is marked as the stable branch.

File: aio/scripts/deploy-to-firebase.sh (hunk 15 above)
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: Hmm, well strictly we are deploying to `aio-staging.firebaseapp.com` it just happens that this is also `next.angular.io`. Happy to change later if that is a problem.

File: aio/scripts/deploy-to-firebase.sh (hunk 16 above)
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: :scream:

This file really should have some kind of unit test :-(

File: aio/scripts/deploy-to-firebase.sh [hunk 18]
```diff
@@ -3,41 +3,83 @@
...
+    tail -n1
+  )
 
-readonly deployEnv=$1
+  # Do not deploy as it is not the latest branch for the given major version
+  if [ $TRAVIS_BRANCH != $mostRecentMinorVersion ]; then
+    echo "Skipping deploy of branch \"${TRAVIS_BRANCH}\" to firebase."
+    echo "There is a more recent branch with the same major version: \"${mostRecentMinorVersion}\""
+    exit 0
+  fi
```
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: So now we will only deploy to `stable` if there is no branch with a the same major and higher minor version. This means that we couldn't have a stable branch of `4.3.x` and then have another branch with `4.4.x`.

File: aio/scripts/deploy-to-firebase.sh (hunk 15 above)
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: Strictly we are deploying stable to `angular-io.firebaseapp.com`, but that didn't stop you from use `https://angular.io` as the deployed URL 😛 

There shouldn't be any practical difference (except maybe that using `next.angular.io` would also verify that the DNS is configured correctly 🎉).

File: aio/scripts/deploy-to-firebase.sh (hunk 18 above)
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: Hm...that sound almost right, but...
What happens when 4.3.x is the stable branch and we want to deploy `4.4.0-beta.1` from the 4.4.x branch?
We need a more elaborate check that only archives branches below the current stable branch.

File: aio/scripts/deploy-to-firebase.test.sh [hunk 19]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env bash
+
+function check {
+  if [[ $1 == $2 ]]; then
+  echo Pass
+  exit 0
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.test.sh]: No indentation 😞

File: aio/scripts/deploy-to-firebase.test.sh [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    `dirname $0`/deploy-to-firebase.sh --dry-run
+  )
+  expected="Skipping deploy because DUMMY_TEST_COMMIT is not the latest commit ($(git ls-remote origin 2.4.x | cut -c1-40))."
+  check "$actual" "$expected"
+)
+
+(
+  echo ===== archive - skip deploy - major version too high
+  actual=$(
+    export TRAVIS_BRANCH=2.1.x
```
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.test.sh]: I might add a test with `TRAVIS_BRANCH.minor > STABLE_BRANCH.minor` (e.g. `TRAVIS_BRANCH=2.4.x`, `STABLE_BRANCH=2.2.x`) just to make it more obvious.
(It is essentally the same check that is in effect for this test, it just makes the intent in that specific corner-case more obvious.)

**Review by gkalpak — APPROVED** [review]
Love the bash script unit test. It's a shame not to run it on CI, though 😞 

LGTM (as soon as Travis is green).

File: scripts/ci/deploy.sh [hunk 21]
```diff
@@ -39,34 +40,14 @@ case ${CI_MODE} in
...
-          # This is upstream master: Deploy to staging.
-          travisFoldStart "deploy.aio.staging"
-            yarn deploy-staging
-          travisFoldEnd "deploy.aio.staging"
-        fi
+        yarn deploy-production
       )
       travisFoldEnd "deploy.aio"
-
     fi
```
Comment (gkalpak) [inline scripts/ci/deploy.sh]: Oh, this `fi` should go (kudos to Travi for catching that).

**Comment by alxhub** [comment]
Removing the merge label as this doesn't have full approval yet. I'll ping it to Igor.

**Review by IgorMinar — APPROVED** [review]

**Comment by vicb** [comment]
conflicts -> blocked
@petebacondarwin could you please rebase ?

**Comment by gkalpak** [comment]
Rebased. (CI failures "inherited" from master.)

**Comment by vicb** [comment]
Thanks for rebasing @gkalpak
